### ISSUE TITLE
assertTrue (et al) can no longer work except for atoms or function calls

### ISSUE THREAD

**Issue by verdammelt** [OdonataResearchLLC/lisp-unit#36]
The commit c752a29ae58126a90d7f8882e01f59043e9f68b8 introduced code (`expand-t-or-f`) which causes the system to now expect only an atom or a function call expression as its argument.

Concretely this prohibits expressions such as `(assert-true (and x y))` from working since `and` is a macro and cannot be with `apply` or `funcall`.

**Comment by namoamitabha** [comment]
I have same issue. Is there someone working on it already?

**Comment by ThomasHermann** [comment]
Check out the branch 'hotfix/assert-true-macro', it reverts to using `expand-assert` instead of `expand-t-or-f`. The consequence of this is that arguments to functions in failed `assert-true`, `assert-false`, and `assert-nil` are not printed.

It is not obvious to me how to print the arguments.

**Comment by verdammelt** [comment]
So when you were using `expand-t-or-f` it displayed the arguments to `form` (the `cdr` of `form`) nicely.  But that didn't work when the function (`car` of `form`) was a macro since the code trying to `apply` it and you can't apply a macro.

Your old code called `expand-asset` if `form` was an atom. What if it did that if `form` was an atom *or* `(car form)` was a function? (I'm doing some googling to figure out how to do that... Will report back if I figure something out. Maybe you already did this googling.)

**Comment by Frechmatz** [comment]
Same problem here. Got around it with a wrapper function but a fix would be appreciated.

**Comment by ThomasHermann** [comment]
Please comment on any updates to this issue, I'll have some time over the holidays to address it.

**Comment by ThomasHermann** [comment]
Pushed a version to hotfix/assert-true-macro that handles macros. It expands the arguments to the macros, but I'm pretty sure the expansion occurs at the wrong time. It's a start, let me know if this works and whether it would be better to simply not expand the macro arguments. Since macro arguments are expanded at the wrong time, they may be more confusing than useful.

**Comment by verdammelt** [comment]
I did a little testing of this change and it looks good. My original use case was to assert the truth of a form such as `(and this that)` and that works well now.

I was able to break it by doing something perverse like: `(lisp-unit:assert-true (and nil (/ 1 0)))`. This macroexpands into a let block in which the `(/ 1 0)` is evaluated and an error is obviously signaled.

Not sure if that is an important case though.
